### ISSUE TITLE
python type annotations too incomplete to be useful?

### ISSUE THREAD

**Issue by dimbleby** [google/or-tools#4120]
python, version 9.9.3963

this release includes some `py.typed` files.  I like to add annotations to my python code and check them strictly, so I am totally in favour of the idea here.

however, the experience of the annotations as they are today is awkward - they generate lots of false positives on simple code.  

Here are the cases that I have found myself having to `# type: ignore`:

```python
from ortools.sat.python import cp_model
model = cp_model.CpModel()
```
result: `error: Call to untyped function "CpModel" in typed context  [no-untyped-call]`

```python
from ortools.sat.python import cp_model
solver = cp_model.CpSolver()
```
result: `error: Call to untyped function "CpSolver" in typed context  [no-untyped-call]`
```python
      status = solver.Solve(model)
      if status in (cp_model.FEASIBLE, cp_model.OPTIMAL):
          pass
```
result: `error: Non-overlapping container check (element type: "CpSolverStatus", container item type: "ValueType")  [comparison-overlap]`
```
  class SolutionPrinter(cp_model.CpSolverSolutionCallback):
      def __init__(self):
         super().__init__()
```
result: `Call to untyped function "__init__" in typed context  [no-untyped-call]`

please consider either improving the annotations to make them more useful, else not publishing `py.typed` until they are more ready for consumption.

**Comment by dimbleby** [comment]
I think three out of those four can be resolved by adding ` -> None` to `__init__()` methods

**Comment by dimbleby** [comment]
Another one:
```python
      model = cp_model.CpModel()  # type: ignore[no-untyped-call]
      foo = model.NewIntVar(1, 26, "foo")
      bar = model.NewIntVar(1, 26, "bar")
      baz = model.NewIntVar(1, 26, "baz")
      quux = model.NewIntVar(1, 26, "quux")
      thud = model.NewIntVar(1, 26, "thud")
     model.Add(quux == foo + bar + baz + thud)
```
result: 
```
foo.py:14: error: No overload variant of "__radd__" of "LinearExpr" matches argument type "SupportsComplex"  [operator]
foo.py:14: note: Possible overload variants:
foo.py:14: note:     def __radd__(self, LinearExpr | IntVar | Integral | int, /) -> LinearExpr | IntVar | Integral | int
foo.py:14: note:     def __radd__(self, LinearExpr | IntVar | Integral | int | Number | float, /) -> LinearExpr | IntVar | Integral | int | Number | float
foo.py:14: note: Left operand is of type "LinearExpr | int | SupportsComplex"
```

**Comment by lperron** [comment]
interesting. Which type checker are you using ? pytype ? pylance ? mypy ?
Pytype does not complain on the cp-sat code.

**Comment by dimbleby** [comment]
mypy

**Comment by dimbleby** [comment]
https://github.com/google/or-tools/issues/4120#issuecomment-1986861312 was not fixed by my pull request: I don't know whether you want to reopen this, or handle it separately, or neither...

**Comment by lperron** [comment]
Sent me the patch

Le sam. 9 mars 2024, 20:10, David Hotham ***@***.***> a
écrit :

**Comment by dimbleby** [comment]
I have no patch for that one, it looks harder than the others!

**Comment by dimbleby** [comment]
possibly `__add__` and `__radd__`  (and `__sub__` and so on) on a `LinearExpr` should always just return another `LinearExpr`?

**Comment by dimbleby** [comment]
ah, because there's a special case at https://github.com/google/or-tools/blob/2319f87902cd269092f06809c2b91ba4f65b56db/ortools/sat/python/cp_model.py#L468-L469 (and also in `__rmul__`)

most pedantically that could return `LinearExpr | Literal[0]`, less precisely perhaps `LinearExpr | int` or similar variation

**Comment by lperron** [comment]
Not sure the Literal[0] is accepted with older python version ( i believe we support 3.8+).

I pushed a fix for __mul__ and __rmul__, as well as missing ValueType.

**Comment by lperron** [comment]
I reverted the ValueType hack. It breaks our code.

Note that while mypy complains heavily, especially with arithmetic and SupportComplex, pytype validates everything I tried flawlessly.

**Comment by lperron** [comment]
pybind11::class_<Domain>(m, "Domain", DOC(operations_research, Domain))
.def_static("all_values", &Domain::AllValues,
DOC(operations_research, Domain, AllValues))
.def_static("from_values", &Domain::FromValues,
DOC(operations_research, Domain, FromValues), arg("values"))
.def_static("from_intervals", &Domain::FromVectorIntervals,
DOC(operations_research, Domain, FromVectorIntervals),
arg("intervals"))
.def_static("from_flat_intervals", &Domain::FromFlatIntervals,
DOC(operations_research, Domain, FromFlatIntervals),
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by proshkin** [comment]
Why are they not static in sorted_interval_list.py?

**Comment by lperron** [comment]
There are no sorted_interval_list.py
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by proshkin** [comment]
# encoding: utf-8
# module ortools.util.python.sorted_interval_list
# from C:\Users\...\ortools\util\python\sorted_interval_list.cp312-win_amd64.pyd
# by generator 1.147
# no doc

# imports
import pybind11_builtins as __pybind11_builtins


class Domain(__pybind11_builtins.pybind11_object):
....


def FromFlatIntervals(self, flat_intervals, p_int=None): # real
signature unknown; restored from __doc__
    """
    FromFlatIntervals(flat_intervals: List[int]) ->
ortools.util.python.sorted_interval_list.Domain

    This method is available in Python, Java and .NET. It allows building
    a Domain object from a flattened list of intervals (long[] in Java and
    .NET, [0, 2, 5, 5, 8, 10] in python).
    """
    pass

**Comment by lperron** [comment]
This is a pyi file.

**Comment by lperron** [comment]
This is a bit g of mypy. It was fixed in 1.9. I need to check it is really fixed.

**Comment by amm0x41r** [comment]
On 9.10, mypy 1.10.0, code such as:
```
from ortools.sat.python import cp_model
...
status = solver.Solve(self.planner_model.model)
if status is cp_model.INFEASIBLE:
    raise RuntimeError("Infeasible")
```
is flagged with this typing error:
`error: Non-overlapping identity check (left operand type: "CpSolverStatus", right operand type: "ValueType")  [comparison-overlap]`
